pilot/pkg/model/jwks_resolver_test.go

```diff
@@ -150,3 +161,110 @@ func TestSetAuthenticationPolicyJwksURIs(t *testing.T) {
 		}
 	}
 }
+
+func TestGetPublicKey(t *testing.T) {
+	r := newJwksResolver(JwtPubKeyExpireDuration, JwtPubKeyRefreshInterval)
+	defer r.Close()
+
+	ms, err := test.NewServer()
+	if err != nil {
+		t.Fatal("failed to create mock server")
+	}
+	if err := ms.Start(); err != nil {
+		t.Fatal("failed to start mock server")
+	}
+	defer func() {
+		_ = ms.Stop()
+	}()
+
+	mockCertURL := ms.URL + "/oauth2/v3/certs"
+
+	cases := []struct {
+		in                string
+		expectedJwtPubkey string
+	}{
+		{
+			in:                mockCertURL,
+			expectedJwtPubkey: test.JwtPubKey1,
+		},
+		{
+			in:                mockCertURL, // Send two same request, mock server is expected to hit only once because of the cache.
+			expectedJwtPubkey: test.JwtPubKey1,
+		},
+	}
+	for _, c := range cases {
+		pk, err := r.GetPublicKey(c.in)
+		if err != nil {
+			t.Errorf("GetPublicKey(%+v) fails: expected no error, got (%v)", c.in, err)
+		}
+		if c.expectedJwtPubkey != pk {
+			t.Errorf("GetPublicKey(%+v): expected (%s), got (%s)", c.in, c.expectedJwtPubkey, pk)
+		}
+	}
+
+	// Verify mock server http://localhost:9999/oauth2/v3/certs was only called once because of the cache.
+	if got, want := ms.PubKeyHitNum, uint64(1); got != want {
+		t.Errorf("Mock server Hit number => expected %d but got %d", want, got)
+	}
+}
+
+func TestJwtPubKeyRefresh(t *testing.T) {
+	r := newJwksResolver(time.Millisecond /*ExpireDuration*/, 2*time.Millisecond /*RefreshInterval*/)
+	defer r.Close()
+
+	ms, err := test.NewServer()
+	if err != nil {
+		t.Fatal("failed to create mock server")
+	}
+	if err := ms.Start(); err != nil {
+		t.Fatal("failed to start mock server")
+	}
+	defer func() {
+		_ = ms.Stop()
+	}()
+
+	mockCertURL := ms.URL + "/oauth2/v3/certs"
+	cases := []struct {
+		in                string
+		expectedJwtPubkey string
+	}{
+		{
+			in: mockCertURL,
+			// Mock server returns JwtPubKey1 for first call.
+			expectedJwtPubkey: test.JwtPubKey1,
+		},
+	}
+	for _, c := range cases {
+		pk, err := r.GetPublicKey(c.in)
+		if err != nil {
+			t.Errorf("GetPublicKey(%+v) fails: expected no error, got (%v)", c.in, err)
+		}
+		if c.expectedJwtPubkey != pk {
+			t.Errorf("GetPublicKey(%+v): expected (%s), got (%s)", c.in, c.expectedJwtPubkey, pk)
+		}
+	}
+
+	// Sleep so that refresher job has chance to run.
+	time.Sleep(10 * time.Millisecond)
```

Comment:
testing with real clock maybe brittle. Do we have a way to use mock clock?

Reply:
It's not straightforward to apply 'mock clock' to time.Ticker that refresh job uses.
One approach I can think of is add  a changedCount to track how many times refresher job detect change happens, and wait changedCount to change in unit test to verify refresh job has run and detect change, fail the test if wait timeout.